backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/extraction/impl/HibernateSearchMultiCollectorManager.java

```diff
@@ -0,0 +1,233 @@
+					collectorManagers.entrySet().iterator().next();
+			return new HibernateSearchMultiCollectorManagerSingle( timeoutManager, entry.getKey(), entry.getValue() );
+		}
+		else {
+			return new HibernateSearchMultiCollectorManagerMulti( timeoutManager, collectorManagers );
+		}
+	}
+
+	private final TimeoutManager timeoutManager;
+
+	@Override
+	public final Collector newCollector() throws IOException {
+		return wrapTimeLimitingCollectorIfNecessary( actualNewCollector() );
+	}
+
+	protected abstract Collector actualNewCollector() throws IOException;
+
+	private Collector wrapTimeLimitingCollectorIfNecessary(Collector collector) {
+		final Deadline deadline = timeoutManager.deadlineOrNull();
+		if ( deadline != null ) {
+			TimeLimitingCollector wrapped = new HibernateSearchTimeLimitingCollector( collector,
+					new LuceneCounterAdapter( timeoutManager.timingSource() ),
+					deadline.checkRemainingTimeMillis() );
+			// The timeout starts from the given baseline, not from when the collector is first used.
+			// This is important because some collectors are applied during a second search.
+			wrapped.setBaseline( timeoutManager.timeoutBaseline() );
+			return wrapped;
+		}
+		return collector;
+	}
+
+	private static Collector unwrapTimeLimitingCollectorIfNecessary(Collector collector) {
+		if ( collector instanceof HibernateSearchTimeLimitingCollector ) {
+			return ( (HibernateSearchTimeLimitingCollector) collector ).originalCollector();
+		}
+		return collector;
+	}
+
+
+	private static class HibernateSearchMultiCollectorManagerMulti extends HibernateSearchMultiCollectorManager {
+		private final Map<CollectorKey<?, ?>, CollectorManager<Collector, ?>> collectorManagers;
+
+		private HibernateSearchMultiCollectorManagerMulti(TimeoutManager timeoutManager,
+				Map<CollectorKey<?, ?>, CollectorManager<Collector, ?>> collectorManagers) {
+			super( timeoutManager );
+			this.collectorManagers = collectorManagers;
+		}
+
+		@Override
+		protected Collector actualNewCollector() throws IOException {
+			Collector[] collectors = new Collector[collectorManagers.size()];
+			int i = 0;
+			for ( CollectorManager<? extends Collector, ?> collectorManager : collectorManagers.values() ) {
+				collectors[i++] = collectorManager.newCollector();
+			}
+			return MultiCollector.wrap( collectors );
+		}
+
+		@Override
+		public MultiCollectedResults reduce(Collection<Collector> collectors) throws IOException {
+			final int size = collectors.size();
+
+			Map<CollectorKey<?, ?>, Object> results = new HashMap<>();
+
+			int i = 0;
+			for ( Map.Entry<CollectorKey<?, ?>, CollectorManager<Collector, ?>> entry : collectorManagers.entrySet() ) {
+				List<Collector> toReduce = new ArrayList<>( size );
+				for ( Collector possiblyWrappedCollector : collectors ) {
+					Collector collector = unwrapTimeLimitingCollectorIfNecessary( possiblyWrappedCollector );
+					if ( collector instanceof MultiCollector ) {
```

Comment:
I think this will always be true, no? Since you override `reduce` in `HibernateSearchMultiCollectorManagerSingle`.

I wouldn't bother with the check; just cast it.